### DISCUSSION TITLE
Const correctness like in C++?

### DISCUSSION THREAD

**Discussion by ghost** [pascalabcnet/pascalabcnet#3016] [General]
Does PascalABC.NET have such a concept like this? On the section `Размерные и ссылочные типы` of the help file, I have seen `var` and `const` being used in function declarations.

**Comment by SunSerega** [comment]
`const` in function parameters in Pascal stops the function from modifying that parameter. But you'll never notice that outside the function, as it only really prevents modifications of the copy of the value, made for that parameter. If the parameter is reference-typed, it only prevents you from setting another reference:
```
type
  t1 = auto class
    x: integer;
  end;
  
procedure p1(const a: t1);
begin
//  a.x := 5; // Error, a is const
  var b := a; // Copy reference to another variable
  b.x := 5; // Now you can set it
end;

begin end.
```
Honestly, this feels like an undertested part of the compiler, because no one uses it. The behaviour of this code in particular makes no sense.

Well, either way, the type system of C++ is amazing, I love digging through STL - trying to understand that mess of templates, `const`/`constexpr`, etc. is like a puzzle game.
But it's also pretty impractically overcomplicated, so I'm pretty sure such a system is never coming to PascalABC.Net, the best to hope for is in-block `const` variable declarations.
